### DISCUSSION TITLE
Add database `afterCreate` pool `init` hook

### DISCUSSION THREAD

**Discussion by rijkvanzanten** [directus/directus#22973] [Feature Requests]
This will allow specific database connection params to be used, like WAL mode on Sqlite, or CockroachDB options like `serial_normalization` etc.

### Discussed in https://github.com/directus/directus/discussions/14153

<div type='discussions-op-text'>

<sup>Originally posted by **xmichaelx** June 28, 2022</sup>
Is Directus using WAL with SQLite backend? If not is it possible to enable it with executing `pragma journal_mode=wal;` somewhere? Can it be configured?

I see that in https://github.com/directus/directus/blob/895100898cd6f0f6121e964e793137059e82fc1b/api/src/database/index.ts
there is line which enables foreign_keys `await run('PRAGMA foreign_keys = ON');` would it be possible to add WAL? or at least do so depending on additional variable?</div>

**Comment by rijkvanzanten** [comment]
Regarding implementation: right now we're only hooking into the `afterCreate` for sqlite and cockroach, as Directus needs to set a couple options to run. We should rework that a bit to have a single `afterCreate` call on knex, with if statements in there for the vendor specific stuff.

**Comment by Vinayak119A** [comment]
Hey, I am Vinayak Kanjker I would like to contribute in this app can you please assign me a task .
This is for my open source contribution class.
please contact me at  vinukanjiker@gmail.com

**Comment by rijkvanzanten** [comment]
@Vinayak119A Wanna give this one a shot? 🙂 

Also what is _open source contribution class_? Very curious to learn more about what class/study/program is teaching that 🙂

**Comment by Vinayak119A** [comment]
@rijkvanzanten Yes sir I would like to give this one a shot . May I contribute?

**Comment by rijkvanzanten** [comment]
For sure! Give it a shot and lemme know how it goes in a PR :)

**Comment by Vinayak119A** [comment]
@rijkvanzanten sir I have tried this code please accept the request 

code to declare hook

```js
// Method 1 via the .init() method
class User extends Model {}
User.init({
  username: DataTypes.STRING,
  mood: {
    type: DataTypes.ENUM,
    values: ['happy', 'sad', 'neutral']
  }
}, {
  hooks: {
    beforeValidate: (user, options) => {
      user.mood = 'happy';
    },
    afterValidate: (user, options) => {
      user.username = 'Toni';
    }
  },
  sequelize
});

// Method 2 via the .addHook() method
User.addHook('beforeValidate', (user, options) => {
  user.mood = 'happy';
});

User.addHook('afterValidate', 'someCustomName', (user, options) => {
  return Promise.reject(new Error("I'm afraid I can't let you do that!"));
});

// Method 3 via the direct method
User.beforeCreate(async (user, options) => {
  const hashedPassword = await hashPassword(user.password);
  user.password = hashedPassword;
});

User.afterValidate('myHookAfter', (user, options) => {
  user.username = 'Toni';
});

Adding database afterCreate pool init hook

router.post('/', function(req, res) {
//node return reply after running the create
  return models.User.create({ username: req.body.username }).then(function() {      
    return res.json({ message: 'New user created' });
  });
});
```

**Comment by rijkvanzanten** [comment]
@Vinayak119A I ehh have no idea what we're looking at above, but it doesn't seem relevant / related to the Directus codebase at all.. 🤷🏻

**Comment by rijkvanzanten** [comment]
Linear: [ENG-235](https://linear.app/directus/issue/ENG-235/add-database-aftercreate-pool-init-hook)

**Comment by alexchudnovsky** [comment]
Hi @rijkvanzanten 

As a temporary work around, there is no reason that WAL could not be enabled directly through something like DB Browser for SQLite right ?

Kind Regards,
Alex

**Comment by rijkvanzanten** [comment]
@alexchudnovsky Hmm good question. I believe some SQLite settings happen on the connection level instead of the database level, so it depends on what you're configuring 👍🏻

**Comment by st17931** [comment]
@rijkvanzanten Is it still open.Can i start working on this as first time contributor in @directus.Else give me some good first issue which is still open.
Thanks.

**Comment by rijkvanzanten** [comment]
Sure! Go for it @st17931

**Comment by joggienl** [comment]
I've been looking into this for a bit and I am wondering what would be a better way forward with this because I think a `afterCreate` init hook will not work properly imo. So below some of my thoughts.

Perhaps good as background: the reason for looking into this is  because I would like a way to hook into SQLite: load extensions (crypto, spatialite) and set some performance-like settings (wal-mode, cache size, etc).

The current way to kinda "hack" these settings into an SQLite database was the snippet from the linked archived discussion https://github.com/directus/directus/discussions/14153#discussioncomment-3044561.

This is not a good solution, especially for SQLite because some pragma's have to be loaded _**for every**_ connection. This is for example true for mod_spatialite. If there are multiple connections in the pool, only the one acquired will get the setting. If the connection get's re-created or if there are more connections: things break.

The `afterCreate` init-hook would be a more proper way to do it. By emitting that event inside the default `afterCreate` it would work for all supported database vendors. But still, this will not work correctly atm: the database is loaded before extensions are loaded. There are multiple parts that require a database connection before extensions even _can_ be loaded. For example, Directus loads settings from the database to load an extension.

So in order to do this properly doing `afterCreate` magic should be done before anything else, because it could affect everything else.

I see two options that could work, I've tested the second option and that one works

1. A new extension type
2. Use a javascript configuration file

## 1: A new extension type

A new extension type that can be loaded before anything else is loaded (an `init` extension?). **Pro**: as extension type it makes it easier for different use-cases we don't know currently and also extensions are easier to distribute. **Con**: due to timing when it should be loaded probably a complex feature (is that worth it if that's true?). Also, it can't use database features like enabled/disabled state.

## 2: Use a javascript configuration file

Add `afterCreate` as option to the javascript configuration file `config.js`. Pro: easy to use, always loads at the proper time. Con: harder to describe how to use it, less easy to distribute.

I've tested this by adding `function` as `ENV_TYPE` specifically for `DB_AFTER_CREATE`. I had to alter `api/database/index.ts` to actually execute the function:

```js
// Part of api/database/index.ts
// Check the pool object inside the knexConfig object

const knexConfig: Knex.Config = {
  client,
  version,
  searchPath,
  connection: connectionString || connectionConfig,
  log: {
    warn: (msg) => {
      // Ignore warnings about returning not being supported in some DBs
      if (msg.startsWith('.returning()')) return;
      if (msg.endsWith('does not currently support RETURNING clause')) return;

      // Ignore warning about MySQL not supporting TRX for DDL
      if (msg.startsWith('Transaction was implicitly committed, do not mix transactions and DDL with MySQL')) return;

      return logger.warn(msg);
    },
    error: (msg) => logger.error(msg),
    deprecate: (msg) => logger.info(msg),
    debug: (msg) => logger.debug(msg),
  },
  pool: {
    // load poolConfig as before
    ...poolConfig,

    // but override the afterCreate
    afterCreate: async function (conn: any, done: any) {
      // check if the config file had the DB_AFTER_CREATE environment variable (and is a function):
      if (connectionConfig['afterCreate'] && typeof connectionConfig['afterCreate'] === 'function') {
        // And if so, execute it
        connectionConfig['afterCreate'](conn);
      }

      // execute the vendor specific afterCreate methods that are defined below
      if (typeof poolConfig.afterCreate === 'function') {
        poolConfig.afterCreate(conn, done);
      }
    },
  },
};
```

Thats it! I can now load my `config.js` file as following:

```js
// config.js

export default function (env) {
  return {
    HOST: '0.0.0.0',
    PORT: '8055',

    SECRET: 'some-fancy-secret',

    DB_CLIENT: 'sqlite3',
    DB_FILENAME: './database/directus.sqlite3',
    DB_POOL__MIN: 0,
    DB_POOL__MAX: 4,

    DB_AFTER_CREATE: async function (conn) {
      const pragmas = [];

      pragmas.push(`PRAGMA busy_timeout = 30000;`);
      pragmas.push(`PRAGMA journal_mode = wal;`);
      pragmas.push(`PRAGMA synchronous = normal;`);
      pragmas.push(`PRAGMA temp_store = memory;`);

      await Promise.all(
        pragmas.map((pragma) => {
          return new Promise((resolve, reject) => {
            conn.run(pragma, (error) => {
              if (error) reject(error);
              else resolve();
            });
          });
        }),
      );

      console.log('settings loaded!');
    },
  };
}
```

Probably multiple suggestions, improvements or other feedback possible here. Curious to find out how we could go forward with this! Obviously happy to contribute with a PR to get it started.

**Reply by joggienl** [reply]
Some more information: I have currently written two extensions that "try" to hook into the database connection:

- [directus-hook-sqlite-perf](https://github.com/joggienl/directus-hook-sqlite-perf)
- [directus-hook-sqlite-spatialite](https://github.com/joggienl/directus-hook-sqlite-spatialite)

I said "try" because it is not working well enough imo. On load, the extension will try to acquire all the connections and apply the required pragmas and loadExtensions needed. They also register to an event from [tarn.js](https://github.com/vincit/tarn.js) `createSuccess` to re-apply the same pragmas or loadExtensions because a connection can be destroyed if not used.

This is not actually a good practice because the `createSuccess` wil run, but Directus can be executing other queries at the same time on this new connection **before** the code in the event-handler had its time to run (it does not block, it is not acquiring the connections!).

A thought that I was having that might work in together with the initial idea of this discusion, the `afterCreate` pool `init` hook, is that an extension would first need to acquire all the connections and just destroy them immediately.  My assumption is that after that the emitting of the init hook WILL actually reach the extension and load all the necessary code. Still: the downside to that is that there wil be code that has used the database without the settings applied.

**Reply by joggienl** [reply]
Cool, I've used my example from changes for `api/src/database/index.ts` and added `await emitter.emitInit('db.after_create', { conn });` before the vendor specific `afterCreate` is being called.

An extension could look like below. The crux here is the `await database.destroy()` and `database.initialize()`. They will make sure the code in the `db.after_create` handler will run.

```js
export default async function ({ init }, { database, logger }) {
  init('db.after_create', async ({ conn }) => {
    const pragmas = [
      'PRAGMA busy_timeout=30000;',
      'PRAGMA journal_mode=wal;',
      'PRAGMA temp_store=memory;',
    ];

    await Promise.all(
      pragmas.map((pragma) => {
        return new Promise((resolve, reject) => {
          conn.run(pragma, (error) => {
            if (error) reject(error);
            else resolve();
          });
        });
      }),
    );

    logger.info('🔥 Settings loaded for connection!');
  });

  await database.destroy();
  await database.initialize();
}

```

**Reply by joggienl** [reply]
Is there a preferred way forward with this? ☺️

**Reply by rijkvanzanten** [reply]
Relying on inline JS for the configuration itself feels dangerous and won't be compatible with any other configuration file format than JS 🤔 I'd rather keep the config itself as static values. That in turn means we'll have to go the new `init` extension route. That feels like a good way to do it to me 🙂

**Reply by joggienl** [reply]
I agree on the config as static in this case, it's more "Directus" imo. So the new `init` route seems the most logical solution indeed.

I am not completely sure how it would look like technically yet, these kinds of extensions should be able to run before the first call for a database connection is being made. So it is not possible to use the database to enable or disable these extensions. Are there more things on top of mind that would work differently?
